### ISSUE TITLE
How to create a signature matrix from isolated cells from bulk data?

### ISSUE THREAD

**Issue by jacobHog** [omnideconv/omnideconv#197]
Hi,

I want to making a signature matrix using isolated cells, where the isolated cells have been sequenced using the bulk approach instead of the single-cell approach. 
I am going to use sample 1-28 from this GEO archive: https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA418779&o=acc_s%3Aa. I am going to align and count the samples, but how do I convert a count-matrix of bulk data into a signature matrix that works as input for DWLS?

Thank you!

**Comment by alex-d13** [comment]
Hi,

Sounds like an interesting approach, this is similar to what first-generation methods did to construct their signature matrices.
If you would just have one sample per cell-type you could just take each cell-type specific sample as column and the intersection of all measured genes as rows, with their expression values as entries in the matrix.
As it seems like you have some replicates per cell type, you can just average the gene expression across all samples of the same cell-type, DWLS also does this when creating its signature matrix (although of course with cells, not samples). DWLS also filters the genes by performing differential expression between each pair of cell-types and only keeping those with a pval < 0.01 and logFC > |0.5|; you could also replicate these steps to reduce the number of genes.

Best,
Alex

**Comment by jacobHog** [comment]
Hi @alex-d13 

Thank for the reply and help.
I am not interested into working with multiple replicates per celltype, I am only going to use sample 1-28 which are 28 different cell types from the same donor. 

Should the signature file consists of feature counts or TPM values?

**Comment by alex-d13** [comment]
Both would actually work, in our tests it did not make a large difference. Also [this](https://www.nature.com/articles/s41467-020-19015-1
) benchmarking paper did not find large differences for DWLS. We used TPMs in our benchmark.

**Comment by alex-d13** [comment]
I dont have an exact number, my suggestion would be about 100 genes. What output do you get when you run DWLS with the verbose=TRUE parameter?

**Comment by jacobHog** [comment]
```
deconvolute_file <- omnideconv::deconvolute(bulk_expression_matrix, dwls_signature_file, method = "dwls", verbose = TRUE)
A model was provided.
DampenedWLS was chosen because multiple values were supplied for "dwls_submethod"
Running DWLS deconvolution module
Error solving the quadratic programming problem. Your dataset might be too small. Run with verbose=TRUE to get the full error message.
Error in quadprog::solve.QP(D/sc, d/sc, A, bzero) : 
  matrix D in quadratic function is not positive definite!
```

**Comment by alex-d13** [comment]
Could you try to add a pseudocount of 1 to your matrix? Maybe you have some 0 values which DWLS does not handle..

**Comment by jacobHog** [comment]
The signature matrix or the count matrix?

**Comment by alex-d13** [comment]
the signature matrix that you used to run DWLS deconvolution

**Comment by jacobHog** [comment]
I do have some 0 value genes for certain cell types (I would expect this, i guess?)  shown here:
![image](https://github.com/user-attachments/assets/0f0f126c-33e1-4da6-b13a-60b12f41ddf4)

But I do not have a full row with only 0's
(checked with any(apply(dwls_signature_file, 1, function(row) all(row == 0))) which returns FALSE)

But you're telling me that I should try to replace all 0 with 1?

**Comment by alex-d13** [comment]
I agree, some 0 values are of course expected. I am just guessing that they could cause the error you are facing.

**Comment by jacobHog** [comment]
It actually works now, thanks!
Should I increase every value in signature matrix by 1 to avoid having 0's in the matrix? If yes, does this have any impact on the results? or what would be the correct approach

**Comment by alex-d13** [comment]
Yes, I would increase every value by 1, not just the 0 cases.
The impact on the results will be close to nothing, because the relative differences between genes remains the same, which is the important quality for deconvolution.

**Comment by jacobHog** [comment]
Thank you so much for helping. It is always a pleasure to get responses when asking for help :)

**Comment by LorenzoMerotto** [comment]
Hello @jacobHog 

I might suggest a different approach. The idea of the signature building is to find marker genes that distinguish the cell types. 
You could try to find the differentially expressed genes using your bulk sequencing, and maybe take the top 20-50 genes.
Once you've done that, you can create the signature matrix by subsetting to keep only the retained genes. 

In addition: 
The error you mentioned (`Error solving the quadratic programming problem`) is related to the bulk samples. My assumption is that it's related to the low number of counts, as this was often happening to me when deconvolving single cell data (interpreting the expression in the single cell as a mini-bulk).
My workaround was deconvolving every sample individually inside a `tryCatch()` context, but I couldn't find a solution to get the fractions for the problematic samples
